pkg/controllers/state/metrics.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"sigs.k8s.io/karpenter/pkg/metrics"
+)
+
+const (
+	stateSubsystem = "state"
+)
+
+var (
+	ClusterStateNodesTotal = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metrics.Namespace,
+			Subsystem: stateSubsystem,
+			Name:      "nodes_total",
+			Help:      "total nodes in karpenter's cluster state",
+		},
+		[]string{},
```

Comment:
Are we going to label this per NodePool? Might be useful to track this back to the number of state nodes per NodePool.

Reply:
thinking about it more, is there an easy way in cluster state to view a relationship of 

nodepool -> nodes? I don't think we have such a mapping in clusterState as of now and I don't think its worth adding for this metric? 
```go
// Cluster maintains cluster state that is often needed but expensive to compute.
type Cluster struct {
	kubeClient    client.Client
	cloudProvider cloudprovider.CloudProvider
	clock         clock.Clock

	mu                        sync.RWMutex
	nodes                     map[string]*StateNode           // provider id -> cached node
	bindings                  map[types.NamespacedName]string // pod namespaced named -> node name
	nodeNameToProviderID      map[string]string               // node name -> provider id
	nodeClaimNameToProviderID map[string]string               // node claim name -> provider id
	daemonSetPods             sync.Map                        // daemonSet -> existing pod
	clusterState     time.Time
	antiAffinityPods sync.Map // pod namespaced name -> *v1.Pod of pods that have required anti affinities
}
```
I do not see any mapping of nodepool -> nodes.


If you feel differently please let me know and I can add it because there may be other utilites to looking up a nodepools StateNodes, just seems rather expensive to represent without a stronger justification.

Reply:
You can get that mapping through labels like we do elsewhere throughout the code. Something like `node.Labels()[v1beta1.LabelNodePoolKey]`

Reply:
I'm realizing now that this suggestion doesn't make a ton of sense since it would be somewhat expensive to compute, at least within the context of the state code. Maybe something we move to the node metrics controller, but I'm fine punting on that bit for now.